#117: Add modAPI to expose selected functions

hecon5 opened the pull request:
Initial version for modAPI to expose commands to referencing database for #117.

hecon5 commented:
The one issue that may occur is that the release type Enum is in a private module (modAddin); the enum should (IMO) be moved to the API to allow this to be referenced outside.

joyfullservice commented on Version Control.accda.src/modules/modAPI.bas:
```diff
@@ -0,0 +1,56 @@
+Option Compare Database
+Option Explicit
+
+'Exposes the release types to referencing files.
+Public Enum eReleaseType
+    Major_Vxx = 0
+    Minor_xVx = 1
+    Build_xxV = 2
+    Same_Version = 3
+End Enum
```
There might be a little confusion here on the purpose of the release type and deployment functions. These are specifically for developing and deploying the add-in itself, not for use outside the add-in application. You wouldn't want to call these from an external database project.

hecon5 replied:
Yes, I think I understood this was to increment the release of the 'calling' database (or, at least I wanted to borrow the capability to keep the functionality consistent). If the "version" isn't for the calling db, then this can be moved over.

hecon5 replied:
Actually, upon review, this appears to do exactly what I was hoping! Looks like it allows me to increment "my" application's information using the same mechanism. Since I'm highly efficient ("laaaazy"), this works sublimely. 

![image](https://user-images.githubusercontent.com/54177882/104353720-3858a480-54d6-11eb-9167-3499355eb673.png)

joyfullservice replied:
That's very cool. 😄 However, from a software design perspective, I would like to keep the add-in versioning functionality protected inside the add-in, with no concern of a user inadvertently altering the calling database. Furthermore, if we change the functionality of this feature in the future, we don't want to inadvertently break other systems that are now depending upon it. (Think encapsulation and visibility)

That being said, you are welcome to copy the relevant code into your project and use as you like.  😄

hecon5 replied:
Fair point; I'll remove that and abscond with the version code; it's very handy, and I like the elegance.

joyfullservice commented on Version Control.accda.src/modules/modAPI.bas:
```diff
@@ -0,0 +1,56 @@
+'Required especially for when users cannot load the addin as an application
+'level addin.
+
+'VCS Functions (Addin specific)
+'These load / unload the VCS Addin, return the installed version, etc.
+Public Function Launch() As Boolean
+    Launch = AddInMenuItemLaunch
+End Function
```
Yes, this what I would envision for the wrapper function. However, keep in mind that this function could become ambiguous when run from a calling database. I would suggest a more specific name like `LaunchVCS`.

hecon5 replied:
Good point; revising.

hecon5 commented:
Updated; also found this article: 
https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/enum-types-defined-in-standard-modules-or-private-classes-cannot-be-used-in-publ

joyfullservice reviewed (changes requested):
I think what I am really looking for here is some simple wrappers for the actual core functionality of the add-in, such as `LaunchVCS` and `ExportSource`. Functions for installing and removing the add-in would be better handled through the normal process of opening the add-in file directly.

From what I understand on the scope of this issue, we are wanting some end points that can trigger the functionality of the add-in, such as exporting source code when the add-in is referenced as a database library rather than installed as an add-in.

Since the general functionality of the add-in at this point assumes that it has been installed as an add-in, we would also need to test to make sure that functions like referencing the default options file still works without errors. Some functionality such as building from source would not work at all in this mode, so that also needs to be taken into consideration.

Stepping back to consider this a little more, this add-in is essentially a tool in the toolbox of a developer. The bridge that enables version control for Microsoft Access database development. The add-in (no longer) requires admin rights to install or remove, installs under the current user's profile (AppData), and works very similar to the built-in access wizards. I am just trying to understand why something like that would not be allowed. It seems like this is part of giving the team the tools required to do the job...

I also want to keep the big picture in mind... I am happy to include a module to add some endpoints if that would help make this workable in your environment, but I guess I am hesitating a bit to do significant refactoring at the core level for something that is more of an edge case that benefits a very small percentage of users.

Let me know what you think. (I am also happy to discuss this offline or by phone/video call if that would be easier.) Thanks!

hecon5 commented:
> 
> 
> I think what I am really looking for here is some simple wrappers for the actual core functionality of the add-in, such as `LaunchVCS` and `ExportSource`. Functions for installing and removing the add-in would be better handled through the normal process of opening the add-in file directly.
> 

Good point; I've updated those accordingly.

> From what I understand on the scope of this issue, we are wanting some end points that can trigger the functionality of the add-in, such as exporting source code when the add-in is referenced as a database library rather than installed as an add-in.
> 

Yes; that's my intent; I'd like to also run the "deploy" (is the deploy to deploy the VCS, or deploy the app?) from a standard call with loose connections to my db. I was looking over your devops tools (useful!), and I don't know if you release accde/r files or the accdb (or something else). We currently only deploy accd(r) files to users (had so many instances of users fat fingering things and breaking things and we've not had an issue since doing this).

> Since the general functionality of the add-in at this point assumes that it has been installed as an add-in, we would also need to test to make sure that functions like referencing the default options file still works without errors. Some functionality such as building from source would not work at all in this mode, so that also needs to be taken into consideration.

Using the VCS I modified, I've actually had some really good luck from having a "template" builder file that contains the VCS loader routines to build from source, because you're right in that having it build from source won't work without the file starting as a referenced file. I'll investigate if this approach works with the addon mode, too.

> 
> Stepping back to consider this a little more, this add-in is essentially a tool in the toolbox of a developer. The bridge that enables version control for Microsoft Access database development. The add-in (no longer) requires admin rights to install or remove, installs under the current user's profile (AppData), and works very similar to the built-in access wizards. I am just trying to understand why something like that would not be allowed. It seems like this is part of giving the team the tools required to do the job...

It has to do with that the "addin" toolbox is not configured/installed by our IT group (reasons are myriad). I am able to "install" the addon (and then it does show up in "Database Tools> Add-Ins > Version Control"), but get the following error when loading Access. Adding the file as a reference makes the link work, and loads VCS forms without issue.
![image](https://user-images.githubusercontent.com/54177882/104349964-a0f15280-54d1-11eb-8153-945a9df35a03.png)

> 
> I also want to keep the big picture in mind... I am happy to include a module to add some endpoints if that would help make this workable in your environment, but I guess I am hesitating a bit to do significant refactoring at the core level for something that is more of an edge case that benefits a very small percentage of users.
> 
> Let me know what you think. (I am also happy to discuss this offline or by phone/video call if that would be easier.) Thanks!

I don't think a significant refactor is required (nor warranted). I think the API calls as shown should suffice (honestly, if it really came down to it, I think just launchVCS, VCSVersion (should that be "VersionVCS" to be consistent?), and exportMySource would suffice for now. It would be nice to expose the build from source (this would allow a template file to autobuild), but that'd be gravy.

joyfullservice commented:
> It has to do with that the "addin" toolbox is not configured/installed by our IT group (reasons are myriad). I am able to "install" the addon (and then it does show up in "Database Tools> Add-Ins > Version Control"), but get the following error when loading Access. Adding the file as a reference makes the link work, and loads VCS forms without issue.
> 
Could you create a new issue for this? I think the problem here is that the location of the add-in is [not trusted](https://www.devhut.net/2017/11/05/access-the-wizard-youve-requested-is-not-installed-or-is-in-a-bad-state/). But the good news is that just like we added the add-in menu items through the registry, we can also [add a trusted location via the registry](https://github.com/joyfullservice/msaccess-devops/blob/44aaaa6921de2b2058923847257fd59f25285588/Application.vbs#L43). I do this automatically in my click-once launchers and it works great in our environment. If we can solve _this_, you will have full functionality of the tool, and we won't have to side-load the reference in each project.

hecon5 commented:
Created a new request; that will be rather helpful to get that straightened out. 

I still think it'd be nice to have API calls to increment the file's utilities (version number, automated export) in my main tool will be really nice; I've managed to have it disconnect upon building for deployment.
Since Access 2013+ does not allow you to compile to an accde via command line, I just have it save a verbatim copy, dump the reference (vcs, and a few other) modules that aren't needed, and I manually compile and save as ACCDE. The ACCDE file is then opened and deployed, so having the addin referenced is quite handy for most cases.

hecon5 reviewed (commented):
Didn't mean to start a review of my self.
